### PR TITLE
[flang][fir] Add FIR Types parser diagnostic tests + cleanup

### PR DESCRIPTION
**Pull request by clementval**
Add diagnostics test for FIR types parsers and remove duplicated diagnostics already provided by MLIR.

### CODE REVIEW

File: flang/test/Fir/invliad-types.fir [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+func private @box2() -> !fir.boxchar<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it6() -> !fir.char<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
```
Comment (schweitzpgi) [inline flang/test/Fir/invliad-types.fir:30]: Interesting consequence here. Either a `?` or an integer is ok, but it looks like the message falls from the control-flow of the parser. (This is merely an observation.)

File: flang/test/Fir/invliad-types.fir [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+// expected-error@+2 {{expected valid keyword}}
+// expected-error@+1 {{expected field type list}}
+func private @dvd4() -> !fir.type<derived4(p:i8){}>
+
+// -----
+
+// expected-error@+2 {{expected valid keyword}}
+// expected-error@+1 {{expected field type list}}
+func private @dvd4() -> !fir.type<derived4(p:i8){}>
```
Comment (schweitzpgi) [inline flang/test/Fir/invliad-types.fir:96]: Note: we may want to allow an empty field list.

Does this need to test the fields and "listiness"? e.g. `{f1}`, `{f1:}`, `{f1:f2}`, `{f1:i32,}`, `{,}`

**Review by schweitzpgi — COMMENTED** [review]
Looking good! Thanks for doing this.

File: flang/test/Fir/invliad-types.fir (hunk 1 above)
Comment (clementval) [reply flang/test/Fir/invliad-types.fir:30]: Yeah the `?` does not trigger a diagnostic and it falls on the next one which is the integer.

File: flang/test/Fir/invliad-types.fir (hunk 2 above)
Comment (clementval) [reply flang/test/Fir/invliad-types.fir]: I just added couple of tests for the field list.

**Review by schweitzpgi — APPROVED** [review]

**Pull request merged**